src/lib/mina_base/party.ml

```diff
@@ -0,0 +1,371 @@
+    Random_oracle.(hash ~init:Hash_prefix.snapp_body (pack_input (to_input t)))
+
+  module Digested = struct
+    type t = Random_oracle.Digest.t
+
+    module Checked = struct
+      type t = Random_oracle.Checked.Digest.t
+    end
+  end
+end
+
+module Predicate = struct
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type t =
+        | Full of Snapp_predicate.Account.Stable.V1.t
+        | Nonce of Account.Nonce.Stable.V1.t
+        | Accept
+      [@@deriving sexp, eq, yojson, hash, compare]
+
+      let to_latest = Fn.id
+    end
+  end]
+end
+
+module Predicated = struct
+  module Poly = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        type ('body, 'predicate) t = {body: 'body; predicate: 'predicate}
+        [@@deriving hlist, sexp, eq, yojson, hash, compare]
+      end
+    end]
+
+    let typ spec =
+      Typ.of_hlistable spec ~var_to_hlist:to_hlist ~var_of_hlist:of_hlist
+        ~value_to_hlist:to_hlist ~value_of_hlist:of_hlist
+  end
+
+  [%%versioned
+  module Stable = struct
+    module V1 = struct
+      type t = (Body.Stable.V1.t, Predicate.Stable.V1.t) Poly.Stable.V1.t
+      [@@deriving sexp, eq, yojson, hash, compare]
+
+      let to_latest = Fn.id
+    end
+  end]
+
+  module Proved = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        type t =
+          (Body.Stable.V1.t, Snapp_predicate.Stable.V1.t) Poly.Stable.V1.t
+        [@@deriving sexp, eq, yojson, hash, compare]
+
+        let to_latest = Fn.id
+      end
+    end]
+
+    module Digested = struct
+      type t = (Body.Digested.t, Snapp_predicate.Digested.t) Poly.t
+
+      module Checked = struct
+        type t = (Body.Digested.Checked.t, Field.Var.t) Poly.t
+      end
+    end
+  end
+
+  module Signed = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        type t = (Body.Stable.V1.t, Account_nonce.Stable.V1.t) Poly.Stable.V1.t
+        [@@deriving sexp, eq, yojson, hash, compare]
+
+        let to_latest = Fn.id
+      end
+    end]
+
+    module Digested = struct
+      type t = (Body.Digested.t, Account_nonce.t) Poly.t
+
+      module Checked = struct
+        type t = (Body.Digested.Checked.t, Account_nonce.Checked.t) Poly.t
+      end
+    end
+
+    module Checked = struct
+      type t = (Body.Checked.t, Account_nonce.Checked.t) Poly.t
+    end
+
+    let typ : (Checked.t, t) Typ.t = Poly.typ [Body.typ (); Account_nonce.typ]
+
+    let dummy : t = {body= Body.dummy; predicate= Account_nonce.zero}
+  end
+
+  module Empty = struct
+    [%%versioned
+    module Stable = struct
+      module V1 = struct
+        type t = (Body.Stable.V1.t, unit) Poly.Stable.V1.t
+        [@@deriving sexp, eq, yojson, hash, compare]
+
+        let to_latest = Fn.id
+      end
+    end]
+
+    let dummy : t = {body= Body.dummy; predicate= ()}
+
+    let create body : t = {body; predicate= ()}
+  end
+end
+
+module Poly = struct
```

Comment:
I worry that we lose quite a lot when the inferred types for values contain `Poly.t`s, e.g. `(_ Foo.Poly.t, _ Bar.Poly.t) Poly.t`, and we don't seem to be using this polymorphism anywhere. It might help readability and improve our ability to construct values if this is a functor instead, e.g.
```ocaml
module Poly (Data : TYPE) (Auth : TYPE) = struct
  type t = {data: Data.t; authorization: Auth.t}
end
```
and then each of the below become
```ocaml
module Foo = struct
  [%%versioned
  module Stable = struct
    module V1 = struct
      type t = Poly(Predicated.Foo.Stable.V1)(Bar.Stable.V1).t =
        {data: Predicated.Foo.Stable.V1.t; authorization: Bar.Stable.V1.t}
      [@@deriving sexp, eq, yojson, hash, compare]

      let to_latest = Fn.id
    end
  end]
end
```

Reply:
I did use the polymorphism once so far in the follow up PR, but I could believe this one helper function is not worth ugly types :smile: 

```ocaml
let account_id (t : ((Body.t, _) Predicated.Poly.t, _) Poly.t) : Account_id.t =
  Account_id.create t.data.body.pk t.data.body.token_id
```